File: go/workflows/book_workflow.go
package workflows

import (
	"fmt"
	"temporal-trip-booking/activities"
	"temporal-trip-booking/app"
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

type BookTripInput struct {
	UserId   string `json:"userId"`
	FlightId string `json:"flightId"`
	HotelId  string `json:"hotelId"`
	CarId    string `json:"carId"`
}

func BookWorkflow(ctx workflow.Context, input BookTripInput) (output string, err error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Book workflow started", "userId", input.UserId)

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	// Create saga to manage compensations
	var saga app.Saga
	defer func() {
		if err != nil {
			disconnectedCtx, _ := workflow.NewDisconnectedContext(ctx)
			saga.Compensate(disconnectedCtx)
		}
		err = nil
	}()

	// Book Flight
	var flight string
	err = workflow.ExecuteActivity(ctx, activities.BookFlight, input.FlightId).Get(ctx, &flight)
	if err != nil {
		logger.Warn("Flight booking failed", "error", err)
		return "Booking cancelled", err
	}
	saga.AddCompensation(activities.UndoBookFlight, input.FlightId)

	// Simulate delay
	logger.Info("Sleeping for 1 second...")
	workflow.Sleep(ctx, time.Second)

	// Book Hotel
	var hotel string
	err = workflow.ExecuteActivity(ctx, activities.BookHotel, input.HotelId).Get(ctx, &hotel)
	if err != nil {
		logger.Warn("Hotel booking failed", "error", err)
		return "Booking cancelled", err
	}
	saga.AddCompensation(activities.UndoBookHotel, input.HotelId)

	// Simulate delay
	logger.Info("Sleeping for 1 second...")
	workflow.Sleep(ctx, time.Second)

	// Book Car
	var car string
	err = workflow.ExecuteActivity(ctx, activities.BookCar, input.CarId).Get(ctx, &car)
	if err != nil {
		logger.Warn("Car booking failed", "error", err)
		return "Booking cancelled", err
	}
	saga.AddCompensation(activities.UndoBookCar, input.CarId)

	// Send Notification
	var notification string
	err = workflow.ExecuteActivity(ctx, activities.NotifyUser, input.UserId).Get(ctx, &notification)

	output = fmt.Sprintf("%s %s %s", flight, hotel, car)
	return output, nil
}

File: python/book_workflow.py
import logging
from datetime import timedelta
import asyncio

from temporalio import workflow
from temporalio.common import RetryPolicy

with workflow.unsafe.imports_passed_through():
    from activities import BookTripInput, book_flight, book_hotel, book_car, notify_user

logging.basicConfig(level=logging.INFO)


@workflow.defn
class BookWorkflow:
    @workflow.run
    async def run(self, input: BookTripInput):
        workflow.logger.info(f"Book workflow started, user_id = {input.userId}")

        activity_args = {
            "start_to_close_timeout": timedelta(seconds=5),
            "retry_policy": RetryPolicy(
                initial_interval=timedelta(seconds=1),
                backoff_coefficient=2,
                maximum_interval=timedelta(seconds=30)
            ),
        }

        # Saga compensations
        compensations = []

        try:
            # Book Flight
            flight = await workflow.execute_activity(book_flight, input, **activity_args)
            compensations.append("undo_book_flight")

            workflow.logger.info("Sleeping for 1 second...")
            await asyncio.sleep(1)

            # Book Hotel
            hotel = await workflow.execute_activity(book_hotel, input, **activity_args)
            compensations.append("undo_book_hotel")

            workflow.logger.info("Sleeping for 1 second...")
            await asyncio.sleep(1)

            # Book Car
            car = await workflow.execute_activity(book_car, input, **activity_args)
            compensations.append("undo_book_car")

            # Send Notification
            await workflow.execute_activity(notify_user, input, **activity_args)

            return f"{flight} {hotel} {car}"

        except Exception as ex:
            workflow.logger.error("Failed to book trip", ex)
            for compensation in reversed(compensations):
                await workflow.execute_activity(
                    compensation,
                    input,
                    **activity_args,
                )

            return "Booking cancelled"
